### ISSUE TITLE
Reduce package size

### ISSUE THREAD

**Issue by evilebottnawi** [Marak/faker.js#499]
Package contains unnecessary directories and files - `locale/.publish`, `.jshintrc` and maybe others. I can do PR if your accept this issue.

**Comment by joeyfigaro** [comment]
Worth mentioning there doesn't seem to be a clear way to import *only* the methods you need. Including the following is 1.2mb raw/768kb compressed:
`internet, name, random, address`

Increasing my lib's size by 768kb to get email addresses, first and last names, random numbers, and addresses is hard to swallow.

**Comment by joeyfigaro** [comment]
![](https://d3vv6lp55qjaqc.cloudfront.net/items/0v2Z043u470b3G0i1M3V/Image%202017-05-19%20at%204.05.15%20PM.public.png?X-CloudApp-Visitor-Id=dcbf0d76441d3d10ce83c320ca952dac&v=63f1acb4)

Going to have to try out chance.js or roll our own helpers after seeing this. Everything highlighted there is from faker.

**Comment by Marak** [comment]
@joeyfigaro 

Not sure what you mean by import here. The faker package contains data for a bunch of APIs as well as localities ( outside of `en` ). If you install the package, you will get all that data.

If you need to reduce your libraries size ( perhaps it's for the browser ), you'll probably want to only include a subset of this data. We've already got custom builds per locality, but not per method. If you need to further optimize the size of the package, you could start removing [require statements](https://github.com/Marak/faker.js/blob/master/lib/index.js#L63) from inside faker lib and re-run the build process.

Not sure what else you would expect or how else this could work. I'm open to suggestion. 

I also am not sure certain the sizes you report are 100% accurate. I am seeing `faker.en.js` browser file at 433kb uncompressed and 306kb minified. That is for all of `en` data. I suspect you may be looking at all data for all localities.

**Comment by Marak** [comment]
@evilebottnawi -

Sounds good, let us know what you can do.

**Comment by evilebottnawi** [comment]
@joeyfigaro you can use tree shaking for avoid unnecessary method in you app, it is not related to `faker.js`, look at `webpack` and `rollup` libraries, thanks!

**Comment by jdat82** [comment]
Sorry guys, I don't really understand how I should import faker to reduce my bundle size. Like @joeyfigaro my bundle has increased by 1.3 MB. With modern imports (typescript), how should I import faker if I only want the uuid() method from the random lib for instance? 

For now I'm only using this method but the project is just starting so there will be a few more.

**Comment by Marak** [comment]
@jdat82 

You can try tree-shaking with something like `webpack` and `rollup` to automatically remove un-used codepaths. You can also manually edit your local faker code to remove modules, methods, and locale data which you don't require ( and rebuild ).

There is probably a way to use `import` the way you want, but I'm not exactly sure how to do it. It might require a few changes to the way we are constructing faker instances.

**Comment by jdat82** [comment]
Indeed, that's my intuition too cause I'm using angular which is using webpack and it doesn't work. You should probably export each method.

**Comment by luanmuniz** [comment]
How about the `locale/.publish` folder. Can we remove it? It's triggering my security alerts because of the eval inside `locale/.publish/scripts/docstrap.lib.js`

**Comment by ghaagsma** [comment]
For anyone curious, I developed a [workaround](https://github.com/miragejs/miragejs/issues/403#issuecomment-611192478) for this issue in a `create-react-app` bootstrapped application that uses `faker` in tandem with `miragejs`. It utilizes code splitting to ensure the modules only required for development environments are never requested in production.

The Mirage team recently published a [release](https://github.com/miragejs/miragejs/releases/tag/v0.1.40) to enable automatic tree shaking for consuming applications. Ideally, `faker` contributors can publish a similar change to provide the same benefit; although, I'm not sure it would be as simple depending on how the modules are exported in this solution.

**Comment by yiga2** [comment]
I have forked https://github.com/pdunn/bq_fake_data to update faker version used, and being able to use new methods like `zipCodeByState()`.
Unfortunately Google BigQuery does not support packages > 1MB for `javascript` user-defined functions, and faker `5.1.0` comes out at 1.3MB.

Is there any way to webpack only for a list of locales  e.g. `en` and `en_US` w/o removing locales from the base package ?

**Comment by Marak** [comment]
I think the main issue for size of the npm package is the fact that all locales are included by default in the package source code.

The only solution I can think of for now is creating separate NPM packages for each locale, and having the main `faker` package only contain `en` by default. We've had a few discussions about this in the past.
